spec/core/ics-003-connection-semantics/UPGRADES.md

```diff
@@ -0,0 +1,407 @@
+it under the status path and restore the original connection.
+
+`statusPath(id) => UpgradeStatus{UPGRADE_ERR}`
+
+A relayer may then submit a `CancelConnectionUpgradeMsg` to the counterparty. Upon receiving this message a chain must verify that the counterparty wrote `UPGRADE_ERR` into its `UpgradeStatus` and if successful, it will restore its original connection as well thus cancelling the upgrade.
+
+If an upgrade message arrives after the specified timeout, then the message MUST NOT execute successfully. Again a relayer may submit a proof of this in a `CancelConnectionUpgradeTimeoutMsg` so that counterparty cancels the upgrade and restores it original connection as well.
+
+```typescript
+function connUpgradeInit(
+    identifier: Identifier,
+    proposedUpgrade: UpgradeConnectionState
+) {
+    // current connection must be OPEN
+    currentConnection = provableStore.get(connectionPath(identifier))
+    abortTransactionUnless(connection.state == OPEN)
+
+    // abort transaction if an unmodifiable field is modified
+    // upgraded connection state must be in `UPGRADE_INIT`
+    // NOTE: Any added fields are by default modifiable.
+    abortTransactionUnless(
+        proposedUpgrade.connection.state == UPGRADE_INIT &&
+        proposedUpgrade.connection.counterpartyConnectionIdentifier == currentConnection.counterpartyConnectionIdentifier &&
+        proposedUpgrade.connection.clientIdentifier == currentConnection.clientIdentifier &&
+        proposedUpgrade.connection.counterpartyClientIdentifier == currentConnection.counterpartyClientIdentifier &&
+    )
+
+    // either timeout height or timestamp must be non-zero
+    abortTransactionUnless(proposedUpgrade.TimeoutHeight != 0 || proposedUpgrade.TimeoutTimestamp != 0)
+
+    upgradeStatus = UpgradeStatus{
+        state: UPGRADE_INIT,
+        timeoutHeight: proposedUpgrade.timeoutHeight,
+        timeoutTimestamp: proposedUpgrade.timeoutTimestamp,
+    }
+
+    provableStore.set(statusPath(identifier), upgradeStatus)
+    provableStore.set(connectionPath(identifier), proposedUpgrade.connection)
+    privateStore.set(restorePath(identifier), currentConnection)
+}
+```
+
+NOTE: It is up to individual implementations how they will provide access-control to the `ConnUpgradeInit` function. E.g. chain governance, permissioned actor, DAO, etc.
+Access control on counterparty should inform choice of timeout values, i.e. timeout value should be large if counterparty's `UpgradeTry` is gated by chain governance.
+
+```typescript
+function connUpgradeTry(
+    identifier: Identifier,
+    proposedUpgrade: UpgradeConnectionState,
+    counterpartyConnection: ConnectionEnd,
+    counterpartyUpgradeStatus: UpgradeStatus,
+    proofConnection: CommitmentProof,
+    proofUpgradeStatus: CommitmentProof,
+    proofHeight: Height
+) {
+    // current connection must be OPEN or UPGRADE_INIT (crossing hellos)
```

Comment:
Actually, the chain not ignoring the TRY messages still needs to handle the crossing hellos logic, so this would not really solve anything. You can just ignore my comment.